Xd 29 updates

pklaschka opened the pull request:
### Contents of this Pull Request
This PR makes changes to
- [x] The typings in the `types` folder
- [ ] Configuration files (`tsconfig.json`, `jsconfig.json`, `.gitignore` etc.)
- [x] The sample files (e.g., `sample.js`)
- [ ] The `README.md` or other files containing documentation
- [ ] Other

### Purpose of this Pull Request
To update the types to the current API status

### Issues resolved by this Pull Request
- fixes #76

pklaschka committed:
— PerPluginStorage, the interface

pklaschka committed:
application module updates

pklaschka committed:
interactions module updates

pklaschka committed:
Selection interface updates

pklaschka committed:
scenegraph updates

pklaschka committed:
Merge remote-tracking branch 'origin/master' into xd-29-updates

# Conflicts:
#	sample.js
#	types/application.d.ts
#	types/index.d.ts
#	types/interactions.d.ts
#	types/scenegraph.d.ts

pklaschka committed:
Resolved merge conflicts

pklaschka commented:
@ericdrobinson Would you mind taking a look at this? Thank you very much in advance 🙂

ericdrobinson commented on types/scenegraph.d.ts:
```diff
@@ -10,6 +10,122 @@ declare interface ScaleFactor {
+ if (jsonString) {  // may be undefined
+    let richObjectCopy = JSON.parse(jsonString);
+    console.log(richObjectCopy.list.length);  // 3
+ }
```
The example markup is totally broken in VSCode (doesn't render correctly). Putting the triple-tick onto its own line above/below the code block and adding stars to each line in the comment appears to have fixed the issue. See:

```suggestion
 * @example 
 * ```
 * // This example shows how to save & retrieve rich JSON data in shared metadata storage.
 * // See below for simpler examples of using individual APIs.
 *  const PLUGIN_ID = "<your manifest's plugin ID here>";
 *  let richObject = {
 *    list: [2, 4, 6],
 *    name: "Hello world"
 *  };
 *  node.sharedPluginData.setItem(PLUGIN_ID, "richData", JSON.stringify(richObject));
 *
 * // Later on...
 * // (This could be in a different plugin, if it passes the original plugin's ID here)
 * let jsonString = node.sharedPluginData.getItem(PLUGIN_ID, "richData");
 * if (jsonString) {  // may be undefined
 *   let richObjectCopy = JSON.parse(jsonString);
 *   console.log(richObjectCopy.list.length);  // 3
 * }
 * ```
```

ericdrobinson commented on types/scenegraph.d.ts:
```diff
@@ -10,6 +10,122 @@ declare interface ScaleFactor {
+declare interface PerPluginStorage {
+    /**
+     * Returns a map where key is plugin ID and value is a nested map containing all the shared metadata for that plugin ID (i.e. the result of calling `getForPluginId()` with that ID).
+     */
```
This is missing part of the [current?] documentation as well as the example.

```suggestion
    /**
     * Returns a map where key is plugin ID and value is a nested map containing all the shared metadata for that plugin ID (i.e. the result of calling `getForPluginId()` with that ID).
     * 
     * This map is a clone of the stored metadata, so modifying it has no effect.
     * 
     * @example
     * ```
     * let allSharedMetadata = node.sharedPluginData.getAll();
     * console.log("Plugin A's 'foo' value:",
     *             allSharedMetadata["A"] && allSharedMetadata["A"].foo);
     * console.log("All of plugin B's shared metadata on this node:",
     *             allSharedMetadata["B"]);
     * console.log("List of plugins storing shared metadata on this node:",
     *             Object.keys(allSharedMetadata));
     * ```
     */
```

ericdrobinson commented on types/scenegraph.d.ts:
```diff
@@ -10,6 +10,122 @@ declare interface ScaleFactor {
+
+
+console.log("Plugin B's shared 'foo' value:",
+    node.sharedPluginData.getForPluginId("B").foo);
```
The example is all screwed up. This appears to fix it:

```suggestion
     * @example
     * ```
     * const MY_PLUGIN_ID = "<your manifest's plugin ID here>";
     * let mySharedMetadata = node.sharedPluginData.getForPluginId(MY_PLUGIN_ID);
     * console.log("My shared 'foo' & 'bar' values:",
     *             mySharedMetadata.foo, mySharedMetadata.bar);
     * 
     * console.log("Plugin B's shared 'foo' value:",
     *             node.sharedPluginData.getForPluginId("B").foo);
     * ```
```

ericdrobinson commented on types/scenegraph.d.ts:
```diff
@@ -10,6 +10,122 @@ declare interface ScaleFactor {
+     *
+     * @example
+     * console.log("All properties stored by plugin A on this node:",
+     node.sharedPluginData.keys("A"));
```
Again, formatting.

```suggestion
     * @example
     * ```
     * console.log("All properties stored by plugin A on this node:",
     *             node.sharedPluginData.keys("A"));
     * ```
```

ericdrobinson commented on types/scenegraph.d.ts:
```diff
@@ -10,6 +10,122 @@ declare interface ScaleFactor {
+     * @example
+     * // These are two different values, stored independently per plugin
+ console.log("Plugin A's 'foo' value:", node.sharedPluginData.getItem("A", "foo"));
+ console.log("Plugin B's 'foo' value:", node.sharedPluginData.getItem("B", "foo"));
```
Formatting.

```suggestion
     * @example
     * ```
     * // These are two different values, stored independently per plugin
     * console.log("Plugin A's 'foo' value:", node.sharedPluginData.getItem("A", "foo"));
     * console.log("Plugin B's 'foo' value:", node.sharedPluginData.getItem("B", "foo"));
     * ```
```

ericdrobinson commented on types/scenegraph.d.ts:
```diff
@@ -10,6 +10,122 @@ declare interface ScaleFactor {
+// ^ ERROR: other plugin's metadata is read-only
+
+console.log(node.sharedPluginData.getItem(MY_PLUGIN_ID, "foo"));  // "42"
+     */
```
Formatting.

```suggestion
     * @example
     * ```
     * const MY_PLUGIN_ID = "<your manifest's plugin ID here>";
     * node.sharedPluginData.setItem(MY_PLUGIN_ID, "foo", "42");
     * 
     * node.sharedPluginData.setItem("other_plugin_id", "foo", "42");
     * // ^ ERROR: other plugin's metadata is read-only
     * 
     * console.log(node.sharedPluginData.getItem(MY_PLUGIN_ID, "foo"));  // "42"
     * ```
```

ericdrobinson commented on types/scenegraph.d.ts:
```diff
@@ -10,6 +10,122 @@ declare interface ScaleFactor {
+console.log(node.sharedPluginData.getItem(MY_PLUGIN_ID, "foo"));  // "42"
+
+node.sharedPluginData.removeItem(MY_PLUGIN_ID, "foo");
+console.log(node.sharedPluginData.getItem(MY_PLUGIN_ID, "foo"));  // undefined
```
Formatting:

```suggestion
     * @example
     * ```
     * const MY_PLUGIN_ID = "<your manifest's plugin ID here>";
     * node.sharedPluginData.setItem(MY_PLUGIN_ID, "foo", "42");
     * console.log(node.sharedPluginData.getItem(MY_PLUGIN_ID, "foo"));  // "42"
     * 
     * node.sharedPluginData.removeItem(MY_PLUGIN_ID, "foo");
     * console.log(node.sharedPluginData.getItem(MY_PLUGIN_ID, "foo"));  // undefined
     * ```
```

ericdrobinson commented on types/scenegraph.d.ts:
```diff
@@ -319,6 +435,15 @@ export class ImageFill {
+     * (**Since**: XD 29)
+     *
+     * A unique identifier for the image asset used by this ImageFill. May be shared by other ImageFills, including those with different cropping, size,
+     rotation, or mirroring. If identical images are imported into XD from separate sources, they may have different `assetId`s however.
```
Formatting. This is a admittedly a bit pedantic.

```suggestion
     * rotation, or mirroring. If identical images are imported into XD from separate sources, they may have different `assetId`s however.
```

ericdrobinson reviewed (changes requested):
The only major issue with this section is the broken `@example`s (at least for VSCode's IntelliSense service). The `*`s before each line within the example block are important, it turns out.

It _also_ appears that the tripple-tick (```) syntax is _optional_. I've left it in with my suggestions to help clear things up.

I will also note that VSCode highlighted two unnecessary `declare` keywords in this file. They do not need to exist in front of individual `class`es as the `module` itself is already `declare`d.

ericdrobinson commented on types/application.d.ts:
```diff
@@ -48,6 +48,17 @@ type RenditionResult = {
+     */
+    name: string,
+    /**
+     * *Semi*-unique document identifier. Duplicating an .xd file on disk will result in two files with the same GUID. Duplicating a document via "Save As" will change its GUID; thus two *cloud* documents will never have the same GUID. The GUID of an Untitled document doesn't change when it is saved for the first time. <br><br>This returns the same value as `scenegraph.root.guid`.
```
Formatting. VSCode doesn't understand inline `<br>`.

```suggestion
     * *Semi*-unique document identifier. Duplicating an .xd file on disk will result in two files with the same GUID. Duplicating a document via "Save As" will change its GUID; thus two *cloud* documents will never have the same GUID. The GUID of an Untitled document doesn't change when it is saved for the first time.
     * 
     * This returns the same value as `scenegraph.root.guid`.
```

ericdrobinson commented on types/application.d.ts:
```diff
@@ -72,3 +83,21 @@ export const appLanguage: string;
+ * >
+ * > _This does **not** indicate the frontmost "active" document window in the XD application._
+ * >
+ * > In XD, each document window loads a separate copy of your plugin. When a given instance of your plugin calls this API, you will always receive information about the document that this instance of the plugin is attached to, even if it's not the active window.
```
Formatting. The [source note](https://adobexdplatform.com/plugin-docs/reference/application.html#applicationactivedocument--documentinfo) doesn't have a line break. (A single line-break in Markdown is ignored. Even though this shows up on two separate lines, it is rendered as a single, flowing paragraph.)

```suggestion
 * > _This does **not** indicate the frontmost "active" document window in the XD application._
 * > In XD, each document window loads a separate copy of your plugin. When a given instance of your plugin calls this API, you will always receive information about the document that this instance of the plugin is attached to, even if it's not the active window.
```

ericdrobinson commented on types/application.d.ts:
````diff
@@ -72,3 +83,21 @@ export const appLanguage: string;
+console.log("Document title: " + documentInfo.name);
+console.log("Document ID: " + documentInfo.guid);
+ ```
+ */
````
While this _**does**_ render correctly in VSCode for some reason, it would be best to get it inline with the rest. I have removed the `js` annotation for two reasons:

1. My previous markup suggestions also elided it.
1. This JavaScript is also valid TypeScript (and vice versa) so it shouldn't matter.

```suggestion
 * @example 
 * ```
 * let application = require("application");
 * let documentInfo = application.activeDocument;
 * console.log("Document title: " + documentInfo.name);
 * console.log("Document ID: " + documentInfo.guid);
 * ```
 */
```

pklaschka replied:
For whatever reason, the best way I've found for including examples is to remove the backticks entirely (tested in VSCode and WebStorm)... I'll do that, for now.

ericdrobinson reviewed (changes requested):
Some formatting suggestions here.

ericdrobinson commented on types/index.d.ts:
```diff
@@ -37,26 +37,37 @@ declare global {
          * Array representing the current selection plus any locked items that the user has attempted to select.
          */
-        itemsIncludingLocked: Array<SceneNode>;
+        readonly itemsIncludingLocked: Array<SceneNode>;
```
Modifying this array will not modify the items. I'd suggest making the array itself Readonly as well. (This suggestion applies to the `.items` property, too.)

Marking the arrays as Readonly will potentially save people from making the mistake of adjusting the "selected nodes array" returned by these getters and think it will affect the actual selection internally (it won't).

```suggestion
        readonly itemsIncludingLocked: ReadonlyArray<SceneNode>;
```

ericdrobinson commented on types/index.d.ts:
```diff
@@ -37,26 +37,37 @@ declare global {
+         * The preferred parent to insert newly added content into. Takes into account the current edit context as well as the "focused artboard" if in the root context.
+         Typically this is the same parent where, for example, XD's shape drawing tools would add items.
+         *
+         * _Selected items are not necessarily all immediate children of the `insertionParent`._ They can be anywhere within the [edit context's](/reference/core/edit-context.md) scope.
```
There's a link in here that should be removed.

Consider doing a pass on link removal as I've seen several others throughout the docs as well.

This seems to be a simple copy-paste issue.

```suggestion
         * _Selected items are not necessarily all immediate children of the `insertionParent`._ They can be anywhere within the edit context's scope.
```

ericdrobinson commented on types/index.d.ts:
```diff
@@ -37,26 +37,37 @@ declare global {
+         * Nodes that are currently selected are always in the current edit context.
+         * @param node
+         */
+        isInEditContext(node: SceneNode): boolean;
```
Not sure if this really matters, but it doesn't include the "Since: XD 28" markup.

pklaschka committed:
Update types/index.d.ts

Co-authored-by: Eric Robinson <eric@sonicbloomgames.com>

pklaschka committed:
Update types/index.d.ts

Co-authored-by: Eric Robinson <eric@sonicbloomgames.com>

ericdrobinson commented on types/scenegraph.d.ts:
```diff
@@ -558,20 +558,22 @@ export interface Bounds {
     height: number;
 }
 
-export interface SceneNode extends SceneNodeClass {}
```
I don't think this is the correct way to go about this change.

Unfortunately, specifying the `BLEND_MODE_*` stuff as properties on the abstract class and revealing that class to the outside world _does_ work in general, but it results in _all subclasses_ suddenly resolving those blend modes:

![image](https://user-images.githubusercontent.com/9142587/82486356-b630f100-9aaa-11ea-95c1-28e7a5b64ed1.png)

I did some quick tests and this does not appear to be how things were implemented as the blend modes are `undefined` outside of the `SceneNode` "class" scope.

ericdrobinson commented:
@pklaschka I've done a first-pass review and hit the major points, I think. Basically:

1. You might want to go through and do a pass on `@example` code to clean it up. I saw other locations throughout the code base with similar issues to those pointed out above.
1. You might want to go through and remove the intra-documentation link formatting. It linkifies in the IntelliSense but shows an error if you click it.
1. Adding `BLEND_MODE_*`s to the `SceneNode` base class definition causes it to appear in **all** subclasses. A different approach should be determined here. You might try resurrecting the `SceneNode` interface (and reverting all that `SceneNodeClass` stuff), add those properties to the `SceneNode` interface, and then declare a `SceneNode` name of a type that _only_ contains the various "static properties" and export that.
    * Note that you do not want to export something like `SceneNode: SceneNode` because it would contain all the properties/methods of the SceneNode class which the actual `SceneNode` object you get in Xd from the `scenegraph` module has nothing public on it but standard Object stuff and the properties.

ericdrobinson commented:
@pklaschka With respect to the third item above, something like this looks to work a treat:

```ts
// Define the shape of the SceneNode "restricted" static properties.
//  NOTE: These properties do not show up on sub-classes so they cannot
//  be a part of the class-inheritance tree.
interface SceneNodeStatic {
    readonly BLEND_MODE_PASSTHROUGH: string;
    readonly BLEND_MODE_NORMAL: string;
    readonly BLEND_MODE_MULTIPLY: string;
    readonly BLEND_MODE_DARKEN: string;
    readonly BLEND_MODE_COLOR_BURN: string;
    readonly BLEND_MODE_LIGHTEN: string;
    readonly BLEND_MODE_SCREEN: string;
    readonly BLEND_MODE_COLOR_DODGE: string;
    readonly BLEND_MODE_OVERLAY: string;
    readonly BLEND_MODE_SOFT_LIGHT: string;
    readonly BLEND_MODE_HARD_LIGHT: string;
    readonly BLEND_MODE_DIFFERENCE: string;
    readonly BLEND_MODE_EXCLUSION: string;
    readonly BLEND_MODE_HUE: string;
    readonly BLEND_MODE_SATURATION: string;
    readonly BLEND_MODE_COLOR: string;
    readonly BLEND_MODE_LUMINOSITY: string;
    readonly FIXED_LEFT: string;
    readonly FIXED_RIGHT: string;
    readonly FIXED_TOP: string;
    readonly FIXED_BOTTOM: string;
    readonly FIXED_BOTH: string;
    readonly POSITION_PROPORTIONAL: string;
    readonly SIZE_FIXED: string;
    readonly SIZE_RESIZES: string;
}

// Export the "name" of SceneNode. This provides access to the "restricted" static
//  properties.
// You could get the "type" of this by using `typeof SceneNode` (especially in type
//  declarations).
export const SceneNode: SceneNodeStatic;

// Export the "type" of SceneNode. This provides access to the "SceneNode" _type_.
export interface SceneNode extends SceneNodeBase {}

/**
 * Base class of all scenegraph nodes. Nodes will always be an instance of some subclass of SceneNode.
 */
declare abstract class SceneNodeBase {
```

_**NOTE:**  In the above, I renamed `SceneNodeClass` to `SceneNodeBase` as it better describes what its purpose is._

During my investigation I saw a few things that don't _quite_ match up with this structure but it was nothing _actually usable_. Just differences in accessible prototype properties/functions. In terms of what is documented by Adobe and what you would want to use these APIs for, the structure I outline in this comment seems to work beautifully.

pklaschka committed:
Fixed a typo

pklaschka commented:
Thank you so much for your input. I'm currently dealing with an ear infection (cf. [my comment in the developer forum](https://forums.adobexdplatform.com/t/hi-im-pablo-who-i-am-in-a-few-words/66/5?u=pklaschka)) and am unable to work on this, but I will look at it as soon as possible. Thank you again 👍.

ericdrobinson commented:
Yikes! Take care and I wish you a swift recovery!

pklaschka committed:
A few first enhancements based on review comments

pklaschka committed:
Merge remote-tracking branch 'origin/xd-29-updates' into xd-29-updates

pklaschka committed:
A few first enhancements based on review comments

pklaschka commented:
> @ericdrobinson:
> You might want to go through and remove the intra-documentation link formatting. It linkifies in the IntelliSense but shows an error if you click it.

Truthfully, I wasn't able to replicate it. VSCode just doesn't link (at all) and WebStorm handled it fine (correctly linking). Do you have any examples where it wasn't working for you (I know it's been quite some time since your comment, sorry about that 😉 )?

pklaschka committed:
A few first enhancements based on review comments

pklaschka committed:
Going back to SceneNode being an interface (as per review comments)

This stops the static constants from getting "implemented" in inheriting classes

ericdrobinson commented:
> Truthfully, I wasn't able to replicate it. VSCode just doesn't link (at all) and WebStorm handled it fine (correctly linking). Do you have any examples where it wasn't working for you (I know it's been quite some time since your comment, sorry about that 😉 )?

Yes. If you attempt to use `scenegraph.getNodeByGUID()` in VSCode, the IntelliSense will appear like so:

![image](https://user-images.githubusercontent.com/9142587/88689752-e0b1b280-d0c8-11ea-97ee-c75e376aa716.png)

Note that "`guid` getter" is shown as a link in VSCode's IntelliSense view. If you _click_ that link, you get the following error:

![image](https://user-images.githubusercontent.com/9142587/88689898-1060ba80-d0c9-11ea-9a19-222512c6da50.png)

The links do not work.

It turns out that intra-API linking in IntelliSense is **not** supported by the TypeScript language server. [This TypeScript issue](https://github.com/microsoft/TypeScript/issues/35524) is tracking interest/development.

While this does [appear to work](https://github.com/microsoft/TypeScript/issues/35524#issuecomment-581330182) in WebStorm (and other JetBrains IDEs), any IDE that relies on the TypeScript Language Server (most non-JetBrains IDEs?) will report errors when you attempt to use this syntax.

That all said, the workaround would be to point to the Adobe XD documentation website. In the case outlined above, this would mean converting this:

    [`guid` getter](#SceneNode-guid)

to this:

    [`guid` getter](https://adobexdplatform.com/plugin-docs/reference/scenegraph.html#SceneNode-guid)

In my tests, this appears to work perfectly. Clicking that link opens the expected dialog in VSCode:

![image](https://user-images.githubusercontent.com/9142587/88691756-42731c00-d0cb-11ea-9041-1ea2208e78f6.png)
